### PR TITLE
Decrecate saveData in favor of saveJson

### PR DESCRIPTION
**Pull request by maartenuni**
Issue #32 describes why the old saveData is suboptimal. This
commit deprecates saveData and adds its replacement function.

### CODE REVIEW

**Comment by tymees** [comment]
Code looks fine to me, same as the reasoning behind it.

The only thing I would say is that technically we don't need to deprecate ``saveData`` and could've changed that function to require a data parameter. 1.0 is going to be backwards incompatible anyway.

That being said, it's a minor thing. Just saying it, as we could reasonably drop the old function.

**Review by tymees — APPROVED** [review]

**Comment by maartenuni** [comment]
> The only thing I would say is that technically we don't need to deprecate `saveData` and could've changed that function to require a data parameter. 1.0 is going to be backwards incompatible anyway.

Yes, you are right about that. However, I also think that saveJson() describes better what it is actually saving. _Data_ is one of the most vague terms in our branch in my opinion.

**Comment by tymees** [comment]
True; one could make the argument that technically you could use that method to sent something other than JSON; in fact, the code itself explicitly sents the JSON with text/plain instead of application/json. Especially as you say the following in the linked issue:

> this will also make the jspsych-uil-utils more general as we can drop the dependency on jsPsych and use the utils in unison with another web framework.

But again, this is me just being pedantic ;) So just ignore me

**Comment by maartenuni** [comment]
I made the final changes, so I hope the current code is clear. Would you please review it once more?

File: jspsych-uil-utils.js [hunk 1]
```diff
@@ -165,6 +167,36 @@ var uil = {};
...
+            access_key = _access_key;
+        }
+
+        let is_online = isOnline(getProtocol(), getHostname());
+
+        if (access_key === KNOWN_INVALID_ID) {
+            let message =
+                `The access_key is "${KNOWN_INVALID_ID}", you should update it.` +
+                "You can find the access_key in globals.js";
+            if (is_online) { // Threat as error when online
```
Comment (bbonf) [inline jspsych-uil-utils.js:189]: typo Threat -> Treat

File: jspsych-uil-error.js [hunk 2]
```diff
@@ -71,6 +71,32 @@ else {
...
+         *                 error dialog.
+         */
+        context.scriptError = function (message) {
+
+            let dialog = _dialog;
+
+            if (_dialog === null) {
+                _dialog = createErrorDialog();
+                dialog = _dialog;
+            }
```
Comment (tymees) [inline jspsych-uil-error.js:90]: I see this bit of code somewhere else also; I'd say we could refactor this into a separate ``getOrCreateDialog`` method?

Could be it's own issue tho

File: jspsych-uil-utils.js [hunk 3]
```diff
@@ -177,6 +179,21 @@ var uil = {};
             // If we do, use that key
             access_key = _access_key;
         }
+
+        let is_online = isOnline(getProtocol(), getHostname());
+
+        if (access_key === KNOWN_INVALID_ID) {
```
Comment (tymees) [inline jspsych-uil-utils.js:185]: Maybe also add a check to validate it's formatted like an UUIDv4? Should be doable with a regex. 

Might help if someone accidentally selects all but the last/first character in the UUID

File: jspsych-uil-utils.js [hunk 4]
```diff
@@ -277,22 +309,72 @@ var uil = {};
...
+     * @deprecated use uil.saveJson() instead.
      */
     context.saveData = function (access_key, acc_server = undefined) {
 
+        if (_access_key) {
+            access_key = _access_key;
+        }
+        else {
+            setAccessKey(access_key);
+            access_key = _access_key;
```
[the remark this answers is not in the document]
Comment (maartenuni) [reply jspsych-uil-utils.js:321]: Do you suggest to use _access_key straight away?

File: jspsych-uil-utils.js (hunk 3 above)
Comment (maartenuni) [reply jspsych-uil-utils.js]: Checks added to see whether the access_key has the format of a uuid. It won't differentiate between a uuid v1,2,3,4 or 5, but just the format, I think that is good enough for our purposes.

File: jspsych-uil-utils.js [hunk 5]
```diff
@@ -182,9 +198,27 @@ var uil = {};
...
         let is_online = isOnline(getProtocol(), getHostname());
 
-        if (access_key === KNOWN_INVALID_ID) {
+        if (!isUUIDFormat(access_key)) {
+            let message =
+                `The access_key ${access_key} is not in a valid format `       +
+                "It should be 5 groups of hexadecimal numbers (0-9 and a-f) "  +
+                "separated by one single '-'."                                 +
+                "The groups should contain 8,4,4,4,12 hexadecimal characters " +
+                "respectively.";
```
Comment (tymees) [inline jspsych-uil-utils.js:207]: While this is 100% factual information, I have my doubts about it's ability to convey meaning to an average user.
I would suggest the following:

The access_key ${access_key} is not in a valid format. Please make sure you copied it correctly into your experiment.
It should be 5 groups of characters (0-9 and a-f) with 8, 4, 4, 4, 12 characters per group respectively.

Comment (maartenuni) [reply jspsych-uil-utils.js:207]: Ok will add that.

File: jspsych-uil-utils.js (hunk 3 above)
Comment (tymees) [reply jspsych-uil-utils.js]: Yeah that's fine; I though of it as an 'oops copy paste error' protection anyway

**Pull request merged**